<!-- badges: start -->
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.15857180.svg)](https://doi.org/10.5281/zenodo.15857180)
[![License](https://img.shields.io/badge/License-GPL_3.0--only-orange.svg)](https://opensource.org/license/gpl-3-0)
<!-- badges: end -->

# Viral Experiment LiP Analysis Workflow
The **Viral Experiment LiP Analysis Workflow** is an adaptable user-friendly notebook for analyzing host-pathogen proteomic datasets using Limited Proteolysis (LiP) standardized experimental sampling methods. This notebook is a supplement to the published dataset collection "Human Host Cellular Response to HCoV-229E Infection Multi-Omics [(ACS-JM-DP2)](https://doi.org/10.25584/PPI/2475744)" evaluating the human host cellular response to wild-type [Human coronavirus strain 229E (HCoV-229E)](https://identifiers.org/genpept:AF304460.1) infection. See [Usage](https://github.com/PNNL-Predictive-Phenomics/Viral-Experiment-LiP-Analysis/blob/main/README.md#usage) for additional requirements information applicable to source code reuse.

**Reference Input Files**

 - Host Sequence Annotation Collection (.FASTA) - Homo sapiens (Human) [UniProt Proteome ID: [UP000005640](https://bioregistry.io/uniprot.proteome:UP000005640)] (version: 20,383 gene entries)
 - Processed Proteomic Datasets (.txt) - A549, MRC5, and primary HAE cell MaxQuant (v.1.6.17.0) [peptide](https://cox-labs.github.io/coxdocs/output_tables.html#all-peptides) and [proteinGroups](https://cox-labs.github.io/coxdocs/output_tables.html#protein-groups) processed raw dataset results files

## Usage
User must specify the search software used to generate the input file in addition to the path of the input files and output files. This workflow currently only supports reference input files generated by [MaxQuant](https://www.maxquant.org) and [MSFragger](https://msfragger.nesvilab.org) search software outputs.

## License
This code is released under a GNU General Public License version 3 (GPL-3.0-only). See our [License](https://github.com/PNNL-Predictive-Phenomics/Viral-Experiment-LiP-Analysis/blob/main/LICENSE) for detailed information.

## Funding
The source code described here was funded by the Predictive Phenomics Science & Technology Initiative, conducted under the Laboratory Directed Research and Development Program, at Pacific Northwest National Laboratory (PNNL). PNNL is a multiprogram national laboratory operated by Battelle for the U.S. Department of Energy Office of Science under Contract No. DE-AC05-76RL0183.

## Citation Guidance
Feng, S. (2025). Predictive Phenomics Host-Pathogen Limited Proteolysis Proteomics Analysis Workflow (1.0.0). Zenodo. https://doi.org/10.5281/zenodo.15857180
